Tools/LambdaTestTool/src/Amazon.Lambda.TestTool.BlazorTester/Services/RuntimeApiDataStore.cs

```diff
@@ -0,0 +1,322 @@
+using System.Collections.ObjectModel;
+using System.Linq;
+using System.Threading;
+
+namespace Amazon.Lambda.TestTool.BlazorTester.Services
+{
+    public interface IRuntimeApiDataStore
+    {
+        EventContainer QueueEvent(string eventBody);
+        
+        IReadOnlyList<IEventContainer> QueuedEvents { get; }
+        
+        IReadOnlyList<IEventContainer> ExecutedEvents { get; }
+
+        void ClearQueued();
+
+        void ClearExecuted();
+
+        void DeleteEvent(string awsRequestId);
+
+
+        IEventContainer ActiveEvent { get; }
+
+        event EventHandler StateChange;
+
+        bool TryActivateEvent(out IEventContainer activeEvent);
+
+        void ReportSuccess(string awsRequestId, string response);
+        void ReportError(string awsRequestId, string errorType, string errorBody);
+    }
+
+    public class RuntimeApiDataStore : IRuntimeApiDataStore
+    {
+        private IList<EventContainer> _queuedEvents = new List<EventContainer>();
+        private IList<EventContainer> _executedEvents = new List<EventContainer>();
+        private int _eventCounter = 1;
+        private object _lock = new object();
+        
+        public event EventHandler StateChange;
+        
+        public EventContainer QueueEvent(string eventBody)
+        {
+            var evnt = new EventContainer(this, _eventCounter++, eventBody);
+            Monitor.Enter(_lock);
+            try
+            {
+                _queuedEvents.Add(evnt);
+                Monitor.PulseAll(_lock);
+            }
+            finally
+            {
+                Monitor.Exit(_lock);
+            }
+            
+            RaiseStateChanged();
+            return evnt;
+        }
+
+        public bool TryActivateEvent(out IEventContainer activeEvent)
+        {
+            activeEvent = null;
+            Monitor.Enter(_lock);
+            try
+            {
+                if (!_queuedEvents.Any())
+                {
+                    return false;
+                }
+
+                var evnt = _queuedEvents[0];
+                _queuedEvents.RemoveAt(0);
+                evnt.EventStatus = IEventContainer.Status.Executing;
+                if (ActiveEvent != null)
+                {
+                    _executedEvents.Add(ActiveEvent as EventContainer);
+                }
+                ActiveEvent = evnt;
+                activeEvent = ActiveEvent;
+                RaiseStateChanged();
+                Monitor.PulseAll(_lock);
+                return true;
+            }
+            finally
+            {
+                Monitor.Exit(_lock);
+            }
+        }
+        
+        public IEventContainer ActiveEvent { get; private set; }
+
+        public IReadOnlyList<IEventContainer> QueuedEvents
+        {
+            get
+            {
+                Monitor.Enter(_lock);
+                try
+                {
+                    return new ReadOnlyCollection<EventContainer>(_queuedEvents); 
```

Comment:
All the places where you are creating a ReadOnlyCollection protected by the lock doesn't seem to be doing what you intend because ReadOnlyCollection constructor doesn't copy the passed collection. So as _queuedEvents is modified so will be the collection the ReadOnlyCollection is wrapping. If you didn't care about the events being added to _queuedEvents under the hood then you wouldn't need the lock.

Did you intend something like this?

`return new ReadOnlyCollection<EventContainer>(_queuedEvents.ToArray());`

Reply:
Added the `ToArray()`